### PR TITLE
feat: use multiple GPUs for inner product proof (PROOF-716)

### PR DESCRIPTION
**Pull request by rnburn**
# Rationale for this change

Use multiple GPUs to compute inner product.

Note: the format check is failing because I think because of a difference between clang-format in the nix environment and the docker image.

# What changes are included in this PR?

* Rewrote inner product GPU driver to use transform algorithm
* Refactored code to use a common workspace structure

# Are these changes tested?

Yes. Reuses existing tests and adds a few others.

### CODE REVIEW

File: sxt/proof/inner_product/generator_fold_kernel.cc [hunk 1]
```diff
@@ -16,49 +16,46 @@
...
+    auto decomposition_device = co_await xendv::make_active_device_viewable(alloc, decomposition);
+    co_return functor{
+        .decomposition_data = decomposition_device.data(),
+        .decomposition_size = static_cast<unsigned>(decomposition_device.size()),
+    };
+  };
+  basit::chunk_options chunk_options{
+      .min_size = 1ull << 9u,
+      .max_size = 1ull << 18u,
+  };
```
Comment (jacobtrombetta) [inline sxt/proof/inner_product/generator_fold_kernel.cc:57]: Are these chunk options optimized or a best guess? I know other chunk options hospitalizations have a comment that they need to undergo testing to find the best values. Perhaps a comment here would help if that is also the case?

File: sxt/proof/inner_product/scalar_fold_kernel.cc [hunk 2]
```diff
@@ -27,68 +28,37 @@
...
+  auto f1 = [m_low, m_high] __device__ __host__(s25t::element & x,
+                                                const s25t::element& y) noexcept {
+    s25o::mul(x, m_low, x);
+    s25o::muladd(x, m_high, y, x);
+  };
+  auto m = scalars.size() - mid;
+  basit::chunk_options chunk_options{
+      .min_size = 2u << 10u,
+      .max_size = 2u << 20u,
+  };
```
Comment (jacobtrombetta) [inline sxt/proof/inner_product/scalar_fold_kernel.cc:52]: I wonder if a comment would help here too? Only if the chuck options need to be tested and optimized.

File: sxt/proof/inner_product/workspace.cc [hunk 3]
```diff
@@ -15,3 +15,34 @@
  * limitations under the License.
  */
 #include "sxt/proof/inner_product/workspace.h"
+
+#include "sxt/base/container/span_utility.h"
+#include "sxt/execution/async/future.h"
```
Comment (jacobtrombetta) [inline sxt/proof/inner_product/workspace.cc:20]: Do we need `future.h` here?

File: sxt/proof/inner_product/workspace.h [hunk 4]
```diff
@@ -16,29 +16,41 @@
  */
 #pragma once
 
+#include <cstddef>
+#include <memory_resource>
+
+#include "sxt/base/container/span.h"
 #include "sxt/execution/async/future_fwd.h"
```
Comment (jacobtrombetta) [inline sxt/proof/inner_product/workspace.h:23]: Do we still need `future_fwd.h`?

Comment (rnburn) [reply sxt/proof/inner_product/workspace.h]: Dropped. Thank you.

File: sxt/proof/inner_product/workspace.cc (hunk 3 above)
Comment (rnburn) [reply sxt/proof/inner_product/workspace.cc]: Dropped. Thank you.

File: sxt/proof/inner_product/scalar_fold_kernel.cc (hunk 2 above)
Comment (rnburn) [reply sxt/proof/inner_product/scalar_fold_kernel.cc:55]: Added commenting.

File: sxt/proof/inner_product/generator_fold_kernel.cc (hunk 1 above)
Comment (rnburn) [reply sxt/proof/inner_product/generator_fold_kernel.cc:60]: Added commenting.

**Review by jacobtrombetta — APPROVED** [review]

**Pull request merged**